File: ios/SentryScopeTier.swift
import Foundation
import Sentry

/// Trims the fields sentry-cocoa attaches to every event down to what each
/// privacy tier promises (docs/sentry-integration-plan.md §9b.3 / §9b.4).
/// Mirror of Android's `TierScopeEventProcessor`.
///
/// The diagnostic (default) tier keeps coarse hardware identity plus OS
/// name/version and app id/version/build; the fingerprint-friendly extras
/// (kernel version, OS build string, app name, locale + timezone, screen
/// metrics) only ship when the user opts into `applicationUsageData`.
///
/// Contexts are filtered through an allowlist rather than deleting a
/// denylist, so a field added by a future SDK version is dropped by default
/// instead of shipped by accident.
enum SentryScopeTier {
    private static let deviceKeysDiagnostic: Set<String> = [
        "manufacturer", "brand", "model", "model_id", "family", "arch", "archs",
        "simulator", "processor_count", "memory_size", "storage_size",
    ]
    private static let deviceKeysUsageExtra: Set<String> = [
        "screen_width_pixels", "screen_height_pixels", "screen_resolution",
        "screen_density", "screen_dpi", "screen_scale", "locale", "timezone",
    ]
    private static let osKeysDiagnostic: Set<String> = ["name", "version"]
    private static let osKeysUsageExtra: Set<String> = ["kernel_version", "build"]
    private static let appKeysDiagnostic: Set<String> = [
        "app_identifier", "app_version", "app_build",
    ]
    private static let appKeysUsageExtra: Set<String> = ["app_name"]

    /// The global tags every native event carries; anything else riding on a
    /// boot transaction is user-shape and dropped at the diagnostic tier.
    private static let bootTransactionTagAllowlist: Set<String> = [
        SentryTags.proc, SentryTags.layer,
    ]

    private static let gb: Int64 = 1 << 30

    /// `beforeSend` hook body. Mutates and returns the event.
    static func trimEvent(_ event: Event, applicationUsageData: Bool) -> Event {
        if var context = event.context {
            trim(&context, key: "device",
                 keep: deviceKeysDiagnostic, usageExtra: deviceKeysUsageExtra,
                 applicationUsageData: applicationUsageData)
            trim(&context, key: "os",
                 keep: osKeysDiagnostic, usageExtra: osKeysUsageExtra,
                 applicationUsageData: applicationUsageData)
            trim(&context, key: "app",
                 keep: appKeysDiagnostic, usageExtra: appKeysUsageExtra,
                 applicationUsageData: applicationUsageData)
            if !applicationUsageData {
                // Locale + timezone are high-entropy fingerprint surfaces.
                context["culture"] = nil
            }
            if var device = context["device"],
               let storage = device["storage_size"] as? NSNumber {
                device["storage_size"] = NSNumber(
                    value: bucketStorageSize(storage.int64Value)
                )
                context["device"] = device
            }
            event.context = context
        }
        // §9b.4: boot transactions carry only phase timings at diagnostic.
        if !applicationUsageData,
           event.type == "transaction",
           event.transaction == "comapeo.boot" {
            event.tags = event.tags?.filter {
                bootTransactionTagAllowlist.contains($0.key)
            }
        }
        return event
    }

    /// `beforeSendSpan` hook body: strip span data (e.g. rootkey `generated`)
    /// from boot phase spans at the diagnostic tier; timings, op, and the bare
    /// phase-name description stay.
    static func trimSpan(_ span: Span, applicationUsageData: Bool) -> Span {
        guard shouldStripBootSpanData(
            operation: span.operation,
            applicationUsageData: applicationUsageData
        ) else { return span }
        for key in span.data.keys {
            span.removeData(key: key)
        }
        return span
    }

    static func shouldStripBootSpanData(
        operation: String,
        applicationUsageData: Bool
    ) -> Bool {
        !applicationUsageData && operation.hasPrefix("boot.")
    }

    /// Round up to a standard marketed size (32/64/…/1024 GB) so exact
    /// formatted-capacity bytes can't fingerprint a device.
    static func bucketStorageSize(_ bytes: Int64) -> Int64 {
        var bucket = 32 * gb
        while bucket < bytes && bucket < 1024 * gb {
            bucket <<= 1
        }
        return bucket
    }

    private static func trim(
        _ context: inout [String: [String: Any]],
        key: String,
        keep: Set<String>,
        usageExtra: Set<String>,
        applicationUsageData: Bool
    ) {
        guard let block = context[key] else { return }
        var allowed = keep
        if applicationUsageData { allowed.formUnion(usageExtra) }
        context[key] = block.filter { allowed.contains($0.key) }
    }
}

File: android/src/main/java/com/comapeo/core/SentryScopeTier.kt
package com.comapeo.core

import io.sentry.EventProcessor
import io.sentry.Hint
import io.sentry.SentryBaseEvent
import io.sentry.SentryEvent
import io.sentry.SentryLevel
import io.sentry.protocol.App
import io.sentry.protocol.Device
import io.sentry.protocol.OperatingSystem
import io.sentry.protocol.SentryTransaction

/**
 * Trims the fields sentry-android attaches to every event down to what each
 * privacy tier promises (docs/sentry-integration-plan.md §9b.3 / §9b.4).
 *
 * The diagnostic (default) tier keeps coarse hardware identity plus OS
 * name/version and app id/version/build; the fingerprint-friendly extras
 * (kernel version, `Build.DISPLAY`, app name, locale + timezone, screen
 * metrics) only ship when the user opts into `applicationUsageData`.
 *
 * Contexts are rebuilt from an allowlist rather than nulling a denylist, so a
 * field added by a future SDK version is dropped by default instead of shipped
 * by accident.
 *
 * Error and fatal events are exempt: full device context is most valuable
 * exactly when something crashed, so they keep the SDK's complete device/os/app
 * scope. Only `culture` (locale + timezone), a fingerprint surface with no
 * debugging value, is still dropped from them at the diagnostic tier.
 */
internal class TierScopeEventProcessor(
    private val applicationUsageData: Boolean,
) : EventProcessor {

    override fun process(event: SentryEvent, hint: Hint): SentryEvent {
        if (isError(event)) {
            if (!applicationUsageData) event.contexts.remove("culture")
        } else {
            trimContexts(event)
        }
        return event
    }

    override fun process(transaction: SentryTransaction, hint: Hint): SentryTransaction {
        trimContexts(transaction)
        if (!applicationUsageData && transaction.transaction == "comapeo.boot") {
            slimBootTransaction(transaction)
        }
        return transaction
    }

    /** Error/fatal captures keep the full native scope — see the class doc. */
    private fun isError(event: SentryEvent): Boolean =
        event.level == SentryLevel.ERROR || event.level == SentryLevel.FATAL

    private fun trimContexts(event: SentryBaseEvent) {
        event.contexts.device?.let { event.contexts.setDevice(allowedDevice(it)) }
        event.contexts.operatingSystem?.let {
            event.contexts.setOperatingSystem(allowedOs(it))
        }
        event.contexts.app?.let { event.contexts.setApp(allowedApp(it)) }
        if (!applicationUsageData) {
            // Locale + timezone are high-entropy fingerprint surfaces.
            event.contexts.remove("culture")
        }
    }

    /**
     * §9b.4: boot transactions stay always-on but carry only phase timings at
     * the diagnostic tier — `boot.kind` reveals foreground/background state
     * and span data (e.g. rootkey `generated`) is install-lifecycle shape.
     */
    private fun slimBootTransaction(transaction: SentryTransaction) {
        transaction.removeTag(SentryTags.BOOT_KIND)
        transaction.spans.forEach { it.setData(null) }
    }

    private fun allowedDevice(device: Device): Device = Device().apply {
        manufacturer = device.manufacturer
        brand = device.brand
        model = device.model
        modelId = device.modelId
        family = device.family
        archs = device.archs
        isSimulator = device.isSimulator
        processorCount = device.processorCount
        memorySize = device.memorySize
        storageSize = device.storageSize?.let { bucketStorageSize(it) }
        if (applicationUsageData) {
            screenWidthPixels = device.screenWidthPixels
            screenHeightPixels = device.screenHeightPixels
            screenDensity = device.screenDensity
            screenDpi = device.screenDpi
            timezone = device.timezone
        }
    }

    private fun allowedOs(os: OperatingSystem): OperatingSystem =
        OperatingSystem().apply {
            name = os.name
            version = os.version
            if (applicationUsageData) {
                kernelVersion = os.kernelVersion
                build = os.build
            }
        }

    private fun allowedApp(app: App): App = App().apply {
        appIdentifier = app.appIdentifier
        appVersion = app.appVersion
        appBuild = app.appBuild
        if (applicationUsageData) {
            appName = app.appName
        }
    }

    companion object {
        private const val GB = 1L shl 30

        /** Round up to a standard marketed size (8/16/32/…/1024 GB) so exact
         *  formatted-capacity bytes can't fingerprint a device. Starts at 8 GB
         *  because devices with 8/16 GB of storage are still in the field. */
        internal fun bucketStorageSize(bytes: Long): Long {
            var bucket = 8 * GB
            while (bucket < bytes && bucket < 1024 * GB) bucket = bucket shl 1
            return bucket
        }
    }
}
